sql/log.cc

```diff
@@ -12863,3 +12936,215 @@ void wsrep_register_binlog_handler(THD *thd, bool trx)
 }
 
 #endif /* WITH_WSREP */
+
+inline bool Binlog_cache_free_flush::check(
+    const MYSQL_BIN_LOG::group_commit_entry *entry) const
+{
+  binlog_cache_data *trx_cache=
+    entry->cache_mngr->get_binlog_cache_data(true);
+  binlog_cache_data *stmt_cache=
+    entry->cache_mngr->get_binlog_cache_data(false);
+
+  if (unlikely(encrypt_binlog ||
+               (entry->using_stmt_cache && entry->using_trx_cache &&
+                !stmt_cache->empty() && !trx_cache->empty())))
+    return false;
+
+  binlog_cache_data *cache_data;
+  if (unlikely(entry->using_stmt_cache && !stmt_cache->empty()))
+    cache_data= stmt_cache;
+  else
+    cache_data= trx_cache;
+
+  /* Do not free flush if total_bytes smaller than limit size. */
+  if (likely(cache_data->get_byte_position() <=
+             opt_binlog_cache_free_flush_threshold))
+    return false;
+
+  /* Do not free flush if reserve space equal to zero. */
+  if (cache_data->file_reserved_bytes() == 0)
+    return false;
+
+  /*
+    Do not free flush if not write tmp file, happened when threshold is
+    smaller than binlog cache size.
+  */
+  if (unlikely(cache_data->cache_log.disk_writes == 0))
+    return false;
+
+  return true;
+}
+
+bool Binlog_cache_free_flush::commit(MYSQL_BIN_LOG::group_commit_entry *entry)
+{
+  bool check_purge= false;
+  THD *thd= entry->thd;
+  binlog_cache_mngr *cache_mngr= entry->cache_mngr;
+  binlog_cache_data *cache_data = cache_mngr->get_binlog_cache_data(true);
+  if (unlikely(!entry->using_trx_cache || cache_data->empty()))
+    cache_data= cache_mngr->get_binlog_cache_data(false);
+
+  /* Sync the temp file before enter log_lock to avoid holding the lock long */
+  if (cache_data->sync_tmp_file())
+    return false;
+
+  thd->wait_for_prior_commit();
+
+  // It will be release by trx_group_commit_leader
+  mysql_mutex_lock(&mysql_bin_log.LOCK_log);
+
+  enum enum_binlog_checksum_alg expected_alg=
+      mysql_bin_log.checksum_alg_reset != BINLOG_CHECKSUM_ALG_UNDEF
+          ? mysql_bin_log.checksum_alg_reset
+          : (enum_binlog_checksum_alg) binlog_checksum_options;
+
+  /*
+    In legacy mode, all events should has a valid position this done by
+    updating log_pos field when writing events from binlog cache to binlog
+    file. Thus rename binlog cache to binlog file is not supported in legacy
+    mode.
+
+    if the cache's checksum alg is not same to the binlog's checksum, it needs
+    to recalculate the checksum. Thus rename binlog cache to binlog file is
+    not supported.
+  */
+  if (!mysql_bin_log.is_open() || opt_binlog_legacy_event_pos ||
+      (expected_alg != cache_data->checksum_opt))
+  {
+    mysql_mutex_unlock(&mysql_bin_log.LOCK_log);
+    return false;
+  }
+
+  m_is_free_flushing= true;
```

Comment:
I originally wanted to avoid to add an argument to other functions. But you are right, LOCK_log is freed earlier that make the logic a little complex. So I followed your suggestion to add a argument to related functions.